File: server/schemas/logs.py
from __future__ import annotations

from datetime import datetime
from typing import Literal

from pydantic import BaseModel

from schemas.base import TimestampRead

class AdminActionLogRead(BaseModel):
    id: str
    # Cycle 15: nullable — supprimer un admin preserve ses entrees du
    # journal (FK SET NULL) : les lignes orphelines portent admin_id NULL.
    admin_id: str | None = None
    action: str
    target_table: str
    target_id: str | None = None
    details: dict | None = None
    created_at: datetime


class AuditLogPageRead(BaseModel):
    """Enveloppe paginee du journal d'audit (cycle 16).

    Avant : liste plate sans total (le front paginait a l'aveugle via
    l'heuristique len == limit). Le total exact permet une vraie
    pagination servable et les compteurs globaux (G).
    """

    items: list[AdminActionLogRead]
    total: int
    limit: int
    offset: int


class AuditStatsRead(BaseModel):
    """Stats du journal d'activite (cycle 16) : compteurs globaux + axe
    par jour (chart H) + top auteurs (chart I).

    `distinct_admins` inclut les lignes orphelines (admin_id NULL apres
    la suppression d'un admin, FK SET NULL cycle 15) : on distingue
    `authors_known` (admin encore en base) de `orphan_rows`.
    """

    total: int
    by_action: dict[str, int] = {}
    par_jour: list[dict] = []
    top_auteurs: list[dict] = []


class ContactStatusUpdate(BaseModel):
    # Cycle 17 : "spam" devient assignable (demande explicite) — le filtre GET
    # l'acceptait deja via les aliases, la creation de statut etait asymetrique.
    status: Literal["new", "read", "replied", "archived", "spam"]


class ContactMessageAdminRead(TimestampRead):
    id: str
    full_name: str
    email: str
    subject_code: str
    subject_label: str
    message: str
    status: str
    ip_hash: str | None = None
    user_agent: str | None = None
    replied_at: datetime | None = None


class EventLogRead(TimestampRead):
    """Evenement technique derive du suivi d'ingestion des offres (scraping).

    Ce backend n'a pas une table generique `logs_evenements`: les evenements
    exploitables les plus proches sont les `OfferIngestionEvent` (un par offre
    traitee lors d'un run de scraping), qu'on expose ici sous une forme
    compatible avec un futur ecran "Journal des erreurs".
    """

    id: str
    module: str = "scraping"
    niveau: str
    action: str
    offer_id: str | None = None
    source_scrape_run_id: str | None = None
    hash_unique: str | None = None
    raw_url: str | None = None
    message: str | None = None


class LogsStatsRead(BaseModel):
    """Stats du journal technique (cycle 17) : evenements d'ingestion +
    messages de contact, pour les compteurs et graphiques de /admin/logs.

    Les compteurs `events_*` sont globaux (toutes dates) sauf mention
    contraire ; les axes `events_par_jour`/`events_par_action` respectent la
    fenetre `days`. Cote contacts, on expose le vocabulaire API (new, read,
    replied, archived, spam) — PAS les valeurs internes stockees.
    """

    # Evenements d'ingestion (toutes dates confondues).
    events_total: int
    events_errors: int
    events_warnings: int
    events_duplicates: int
    # Axes sur la fenetre demandee.
    events_par_jour: list[dict] = []
    events_par_action: list[dict] = []
    # Messages de contact (soft-deleted exclus, comme la liste).
    contacts_total: int
    contacts_par_statut: dict[str, int] = {}
    # fenetre effective utilisee pour les axes (echo, aide au debug).
    days: int

File: client/src/api/admin/logs.js
import adminApi from "./axiosAdmin";
import { cleanParams } from "../utils";

/* Journaux : audit des actions admin, evenements de scraping, messages de contact. */

const API_URL = "/api/admin/logs";

/**
 * GET /api/admin/logs/audit — historique des actions admin.
 * Cycle 16 : enveloppe PAGINEE { items, total, limit, offset } — le total
 * est exact SOUS FILTRES (pagination honnete, plus d'heuristique).
 * @param {Object} params admin_id, action ("creation"|"modification"|"suppression"|
 *   "envoi"|"connexion"|"scraping"), target_table, limit (1-200, defaut 50), offset (>= 0)
 */
const getAuditLogs = async (params = {}, { signal } = {}) => {
  const response = await adminApi.get(`${API_URL}/audit`, { params: cleanParams(params), signal });
  return response.data;
};

/**
 * GET /api/admin/logs/audit/stats — stats du journal (cycle 16).
 * @param {Object} params days (1-365, defaut 30) : fenetre de par_jour/top_auteurs.
 *   Réponse : { total (global), by_action (global), par_jour [{jour,total}],
 *   top_auteurs [{admin_id,nom,email,total}] } — orphelins = « Admin supprimé ».
 */
const getAuditStats = async (params = {}, { signal } = {}) => {
  const response = await adminApi.get(`${API_URL}/audit/stats`, { params: cleanParams(params), signal });
  return response.data;
};

/**
 * GET /api/admin/logs/stats — stats du journal technique (cycle 17).
 * @param {Object} params days (1-365, defaut 30) : fenetre des axes.
 *   Réponse : { events_total, events_errors, events_warnings, events_duplicates,
 *   events_par_jour [{jour,total,par_action}], events_par_action [{action,total}],
 *   contacts_total, contacts_par_statut {new,read,replied,archived,spam} (vocabulaire API), days }
 */
const getLogsStats = async (params = {}, { signal } = {}) => {
  const response = await adminApi.get(`${API_URL}/stats`, { params: cleanParams(params), signal });
  return response.data;
};

/**
 * GET /api/admin/logs/events — evenements d'ingestion.
 * @param {Object} params module ("scraping" seul disponible), level ("info"|"warning"|"error"),
 *   source_id, limit (1-200, defaut 50), offset (>= 0)
 */
const getEventLogs = async (params = {}, { signal } = {}) => {
  const response = await adminApi.get(`${API_URL}/events`, { params: cleanParams(params), signal });
  return response.data;
};

/**
 * GET /api/admin/logs/contacts — messages de contact.
 * @param {Object} params status ("new"|"read"|"replied"|"archived"|"spam"), limit (1-100, defaut 20), offset
 */
const getContactMessages = async (params = {}, { signal } = {}) => {
  const response = await adminApi.get(`${API_URL}/contacts`, { params: cleanParams(params), signal });
  return response.data;
};

/**
 * PATCH /api/admin/logs/contacts/{contact_id}/status — body { status: "new"|"read"|"replied"|"archived"|"spam" }.
 * Cycle 17 : "spam" est desormais assignable. La réponse renvoie le statut
 * STOCKE (valeurs internes : "in_progress" pour read, "closed" pour archived,
 * "spam" identique) — le front affiche via la map STATUTS_CONTACT.
 */
const updateContactStatus = async (contactId, status, { signal } = {}) => {
  const response = await adminApi.patch(
    `${API_URL}/contacts/${encodeURIComponent(contactId)}/status`,
    { status },
    { signal },
  );
  return response.data;
};

export { getAuditLogs, getAuditStats, getLogsStats, getEventLogs, getContactMessages, updateContactStatus };

export default { getAuditLogs, getAuditStats, getLogsStats, getEventLogs, getContactMessages, updateContactStatus };
